README.md
CatBoost Rust Package
======================
### Cargo Package
Cargo package can be found [here](https://crates.io/crates/catboost-rs)


### Historical Context
* This started off as an [attempt](https://github.com/catboost/catboost/pull/2161) to publish the catboost bindings to cargo
* Eventually decided to unofficially split out the code and maintain separate rust bindings, similar to `onnxruntime-rs`
* Note that this is voluntarily maintained and not yet endorsed by the official catboost team

### Differences versus official [rust package](https://github.com/catboost/catboost/tree/master/catboost/rust-package)
* Necessary changes made to publish crate instead of referencing the git repo, which is pretty big
* The official `catboost-sys` one attempts to rebuild the shared library, whereas this one downloads it from the github release page.
* The `build.rs` script is with assumption `libcatboost` shared library already downloaded separately.
* Also marked the Model as `Send` so that it can be used across threads, due to the documentation stating it's thread safe. Note that this is not yet extensively tested though.
* As of the present the catboost version is hardcoded, it is currently 1.0.6.

### If you need dependencies for bindgen
```
apt-get install -y curl build-essential pkg-config libssl-dev libclang-dev clang cmake
```

### Install libcatboost
* Download catboost binary from `https://github.com/catboost/catboost/releases/tag/v1.0.6`.
  * If you are using Linux, download `libcatboostmodel.so`
  * If you are using MacOS, download `libcatboostmodel.dylib`
  * Place the file in `/usr/lib/`
  * Create a soft link to `x.x.1`, e.g. `sudo ln -s libcatboostmodel.so libcatboostmodel.so.1` 

### Basic usage example
1. Add a dependency to your Cargo.toml:
```
[dependencies]
catboost-rs = "0.1.6"
```
2. To use catboost, it assumes the shared libraries are available. You will need to download the shared library from the official [releases page](https://github.com/catboost/catboost/releases). If you are using linux, download `libcatboostmodel.so`. If you are using Mac, download `libcatboostmodel.dylib`. As of the present, only version 1.0.6 is supported.
3. Move these libraries to `/usr/lib` 
4. Now you can apply pretrained model in your code:
```rust
// Bring catboost module into the scope
use catboost_rs as catboost;

fn main() {
    // Load the trained model
    let model = catboost::Model::load("tmp/model.bin").unwrap();

    println!("Number of cat features {}", model.get_cat_features_count());
    println!("Number of float features {}", model.get_float_features_count());

    // Apply the model
    let prediction = model
        .calc_model_prediction(
            vec![
                vec![-10.0, 5.0, 753.0],
                vec![30.0, 1.0, 760.0],
                vec![40.0, 0.1, 705.0],
            ],
            vec![
                vec![String::from("north")],
                vec![String::from("south")],
                vec![String::from("south")],
            ],
        )
        .unwrap();
    println!("Prediction {:?}", prediction);
}
```

### Documentation
Run `cargo doc --open` in `catboost/rust-package` directory.

### Tests
Run `cargo test` in `catboost/rust-package` directory.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]


## [0.1.5] - 2022-10-15
### Changed
* Updated `Cargo.toml` to point to right repository.

## [0.1.4] - 2022-10-15
### Changed
* Moved code to it's own repository, splitting it off from the original forked code

## [0.1.3] - 2022-10-15
### Changed
* Mark `Model` as sync as it should be thread safe according to the documentation

catboost-sys/build.rs
use anyhow::{Context, Result};
use std::env;
use std::fs::create_dir_all;

use std::path::PathBuf;

fn main() -> Result<()> {
    let target = env::var("TARGET").unwrap();

    let out_dir =
        PathBuf::from(env::var("OUT_DIR").context("Couldn't get OUT_DIR from env variable")?);

    let build_dir = PathBuf::from(
        env::var("CARGO_MANIFEST_DIR")
            .context("Couldn't get CARGO_MANIFEST_DIR from env variable")?,
    );

    let out_model_interface_dir = out_dir.join("catboost/libs/model_interface");

    create_dir_all(&out_model_interface_dir)
        .context("Couldn't create model interface directory")?;

    let c_bindings_model_interface_path = build_dir.join("c_bindings/model_interface");

    let c_bindings_model_interface =
        c_bindings_model_interface_path
            .canonicalize()
            .context(format!(
                "Failed to find model_interface directory with c bindings: {:?}",
                c_bindings_model_interface_path
            ))?;

    let bindings = bindgen::Builder::default()
        .header("wrapper.h")
        .clang_arg(format!("-I{}", c_bindings_model_interface.display()))
        .size_t_is_usize(true)
        .rustfmt_bindings(true)
        .generate()
        .expect("Unable to generate bindings.");

    bindings
        .write_to_file(out_dir.join("bindings.rs"))
        .context("Couldn't write bindings.")?;

    if target.contains("apple") {
        println!("cargo:rustc-link-lib=c++");
    } else {
        println!("cargo:rustc-link-lib=stdc++");
    }
    println!("cargo:rustc-link-lib=dylib=catboostmodel");

    return Ok(());
}


catboost-sys/src/lib.rs
#![allow(non_upper_case_globals)]
#![allow(non_camel_case_types)]
#![allow(non_snake_case)]

include!(concat!(env!("OUT_DIR"), "/bindings.rs"));

#[cfg(test)]
mod tests {
    use super::*;
    use std::ffi::CStr;

    #[test]
    fn it_works() {
        let model_handle = unsafe { ModelCalcerCreate() };
        let ret_val = unsafe {
            LoadFullModelFromFile(
                model_handle,
                std::ffi::CString::new("files/model.bin").unwrap().as_ptr(),
            )
        };
        if !ret_val {
            let c_str = unsafe { CStr::from_ptr(GetErrorString()) };
            let str_slice = c_str.to_str().unwrap();
            panic!(str_slice);
        }

        let tree_count = unsafe { GetTreeCount(model_handle) };
        assert_eq!(tree_count, 1000);

        let float_features_count = unsafe { GetFloatFeaturesCount(model_handle) };
        assert_eq!(float_features_count, 3);
        let cat_features_count = unsafe { GetCatFeaturesCount(model_handle) };
        assert_eq!(cat_features_count, 1);

        unsafe { ModelCalcerDelete(model_handle) };
    }

    #[test]
    fn it_works_buffer() {
        let buffer = read_fast("files/model.bin").unwrap();
        let model_handle = unsafe { ModelCalcerCreate() };
        let ret_val = unsafe {
            LoadFullModelFromBuffer(
                model_handle,
                buffer.as_ptr() as *const std::os::raw::c_void,
                buffer.len(),
            )
        };
        if !ret_val {
            let c_str = unsafe { CStr::from_ptr(GetErrorString()) };
            let str_slice = c_str.to_str().unwrap();
            panic!(str_slice);
        }

        let tree_count = unsafe { GetTreeCount(model_handle) };
        assert_eq!(tree_count, 1000);

        let float_features_count = unsafe { GetFloatFeaturesCount(model_handle) };
        assert_eq!(float_features_count, 3);
        let cat_features_count = unsafe { GetCatFeaturesCount(model_handle) };
        assert_eq!(cat_features_count, 1);

        unsafe { ModelCalcerDelete(model_handle) };
    }
    use std::io::Read;
    fn read_fast<P: AsRef<std::path::Path>>(path: P) -> std::io::Result<Vec<u8>> {
        let mut file = std::fs::File::open(path)?;
        let meta = file.metadata()?;
        let size = meta.len() as usize;
        let mut data = Vec::with_capacity(size);
        data.resize(size, 0);
        file.read_exact(&mut data)?;
        Ok(data)
    }
}


catboost-rs/src/model.rs
use crate::error::{CatBoostError, CatBoostResult};
use catboost_sys;
use std::ffi::CString;
use std::os::unix::ffi::OsStrExt;
use std::path::Path;

pub struct Model {
    handle: *mut catboost_sys::ModelCalcerHandle,
}

impl Model {
    fn new() -> Self {
        let model_handle = unsafe { catboost_sys::ModelCalcerCreate() };
        Model {
            handle: model_handle,
        }
    }

    /// Load a model from a file
    pub fn load<P: AsRef<Path>>(path: P) -> CatBoostResult<Self> {
        let model = Model::new();
        let path_c_str = CString::new(path.as_ref().as_os_str().as_bytes()).unwrap();
        CatBoostError::check_return_value(unsafe {
            catboost_sys::LoadFullModelFromFile(model.handle, path_c_str.as_ptr())
        })?;
        Ok(model)
    }

    /// Load a model from a buffer
    pub fn load_buffer<P: AsRef<Vec<u8>>>(buffer: P) -> CatBoostResult<Self> {
        let model = Model::new();
        CatBoostError::check_return_value(unsafe {
            catboost_sys::LoadFullModelFromBuffer(
                model.handle,
                buffer.as_ref().as_ptr() as *const std::os::raw::c_void,
                buffer.as_ref().len(),
            )
        })?;
        Ok(model)
    }

    /// Calculate raw model predictions on float features and string categorical feature values
    pub fn calc_model_prediction(
        &self,
        float_features: Vec<Vec<f32>>,
        cat_features: Vec<Vec<String>>,
    ) -> CatBoostResult<Vec<f64>> {
        let mut float_features_ptr = float_features
            .iter()
            .map(|x| x.as_ptr())
            .collect::<Vec<_>>();

        let hashed_cat_features = cat_features
            .iter()
            .map(|doc_cat_features| {
                doc_cat_features
                    .iter()
                    .map(|cat_feature| unsafe {
                        catboost_sys::GetStringCatFeatureHash(
                            cat_feature.as_ptr() as *const std::os::raw::c_char,
                            cat_feature.len(),
                        )
                    })
                    .collect::<Vec<_>>()
            })
            .collect::<Vec<_>>();

        let mut hashed_cat_features_ptr = hashed_cat_features
            .iter()
            .map(|x| x.as_ptr())
            .collect::<Vec<_>>();

        let mut prediction = vec![0.0; float_features.len()];
        CatBoostError::check_return_value(unsafe {
            catboost_sys::CalcModelPredictionWithHashedCatFeatures(
                self.handle,
                float_features.len(),
                float_features_ptr.as_mut_ptr(),
                float_features[0].len(),
                hashed_cat_features_ptr.as_mut_ptr(),
                cat_features[0].len(),
                prediction.as_mut_ptr(),
                prediction.len(),
            )
        })?;
        Ok(prediction)
    }

    /// Apply sigmoid to get predict probability
    // https://catboost.ai/en/docs/concepts/output-data_model-value-output#classification
    pub fn calc_predict_proba(
        &self,
        float_features: Vec<Vec<f32>>,
        cat_features: Vec<Vec<String>>,
    ) -> CatBoostResult<Vec<f64>> {
        let raw_results = self.calc_model_prediction(float_features, cat_features)?;
        let probabilities = raw_results.into_iter().map(sigmoid).collect();
        Ok(probabilities)
    }

    /// Get expected float feature count for model
    pub fn get_float_features_count(&self) -> usize {
        unsafe { catboost_sys::GetFloatFeaturesCount(self.handle) }
    }

    /// Get expected categorical feature count for model
    pub fn get_cat_features_count(&self) -> usize {
        unsafe { catboost_sys::GetCatFeaturesCount(self.handle) }
    }

    /// Get number of trees in model
    pub fn get_tree_count(&self) -> usize {
        unsafe { catboost_sys::GetTreeCount(self.handle) }
    }

    /// Get number of dimensions in model
    pub fn get_dimensions_count(&self) -> usize {
        unsafe { catboost_sys::GetDimensionsCount(self.handle) }
    }
}

impl Drop for Model {
    fn drop(&mut self) {
        unsafe { catboost_sys::ModelCalcerDelete(self.handle) };
    }
}

// Should be thread safe as stated here: https://github.com/catboost/catboost/issues/272
unsafe impl Send for Model {}

unsafe impl Sync for Model {}

fn sigmoid(x: f64) -> f64 {
    1. / (1. + (-x).exp())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn load_model() {
        let model = Model::load("files/model.bin");
        assert!(model.is_ok());
    }

    #[test]
    fn load_model_buffer() {
        let buffer: Vec<u8> = read_fast("files/model.bin").unwrap();
        let model = Model::load_buffer(buffer);
        assert!(model.is_ok());
    }

    #[test]
    fn calc_prediction() {
        let model = Model::load("files/model.bin").unwrap();
        let prediction = model
            .calc_model_prediction(
                vec![
                    vec![-10.0, 5.0, 753.0],
                    vec![30.0, 1.0, 760.0],
                    vec![40.0, 0.1, 705.0],
                ],
                vec![
                    vec![String::from("north")],
                    vec![String::from("south")],
                    vec![String::from("south")],
                ],
            )
            .unwrap();

        assert_eq!(prediction[0], 0.9980003729960197);
        assert_eq!(prediction[1], 0.00249414628534181);
        assert_eq!(prediction[2], -0.0013677527881450977);
    }

    #[test]
    fn get_model_stats() {
        let model = Model::load("files/model.bin").unwrap();

        assert_eq!(model.get_cat_features_count(), 1);
        assert_eq!(model.get_float_features_count(), 3);
        assert_eq!(model.get_tree_count(), 1000);
        assert_eq!(model.get_dimensions_count(), 1);
    }

    use std::io::Read;
    fn read_fast<P: AsRef<std::path::Path>>(path: P) -> std::io::Result<Vec<u8>> {
        let mut file = std::fs::File::open(path)?;
        let meta = file.metadata()?;
        let size = meta.len() as usize;
        let mut data = Vec::with_capacity(size);
        data.resize(size, 0);
        file.read_exact(&mut data)?;
        Ok(data)
    }
}
